# Data Preprocessing

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# 1. Data Loading

In [ ]:
df=pd.read_csv("/content/drive/MyDrive/IPL_Data_Analysis/att_0_1778303821_c3a907.csv")
df.head()

/tmp/ipykernel_1291/4067506012.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv("/content/drive/MyDrive/IPL_Data_Analysis/att_0_1778303821_c3a907.csv")


,match_id,date,season,event,venue,city,team1,team2,toss_winner,toss_decision,...,non_striker,runs_batter,runs_extras,runs_total,extras_wides,extras_noballs,extras_byes,extras_legbyes,wicket_kind,wicket_player_out
0,1082591,2017-04-05,2017,Indian Premier League,"Rajiv Gandhi International Stadium, Uppal",Hyderabad,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,...,S Dhawan,0,0,0,0,0,0,0,NaN,NaN
1,1082591,2017-04-05,2017,Indian Premier League,"Rajiv Gandhi International Stadium, Uppal",Hyderabad,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,...,S Dhawan,0,0,0,0,0,0,0,NaN,NaN
2,1082591,2017-04-05,2017,Indian Premier League,"Rajiv Gandhi International Stadium, Uppal",Hyderabad,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,...,S Dhawan,4,0,4,0,0,0,0,NaN,NaN
3,1082591,2017-04-05,2017,Indian Premier League,"Rajiv Gandhi International Stadium, Uppal",Hyderabad,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,...,S Dhawan,0,0,0,0,0,0,0,NaN,NaN
4,1082591,2017-04-05,2017,Indian Premier League,"Rajiv Gandhi International Stadium, Uppal",Hyderabad,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,...,S Dhawan,0,2,2,2,0,0,0,NaN,NaN


# 2. Data Understanding

### Data Size - Rows and Columns

In [ ]:
df.shape

(289673, 30)

### Data Types

In [ ]:
df.dtypes

,0
match_id,int64
date,object
season,object
event,object
venue,object
city,object
team1,object
team2,object
toss_winner,object
toss_decision,object


### Unique Values in Some Columns

In [ ]:
for column in df.columns:
    if df[column].dtype == 'object':
        print(f"Unique values in column '{column}':\n{df[column].unique()}\n")

Unique values in column 'date':
['2017-04-05' '2017-04-06' '2017-04-07' '2017-04-08' '2017-04-09'
 '2017-04-10' '2017-04-11' '2017-04-12' '2017-04-13' '2017-04-14'
 '2017-04-15' '2017-04-16' '2017-04-17' '2017-04-18' '2017-04-19'
 '2017-04-20' '2017-04-21' '2017-04-22' '2017-04-23' '2017-04-24'
 '2017-04-26' '2017-04-27' '2017-04-28' '2017-04-29' '2017-04-30'
 '2017-05-01' '2017-05-02' '2017-05-03' '2017-05-04' '2017-05-05'
 '2017-05-06' '2017-05-07' '2017-05-08' '2017-05-09' '2017-05-10'
 '2017-05-11' '2017-05-12' '2017-05-13' '2017-05-14' '2017-05-16'
 '2017-05-17' '2017-05-19' '2017-05-21' '2018-04-07' '2018-04-08'
 '2018-04-09' '2018-04-10' '2018-04-11' '2018-04-12' '2018-04-13'
 '2018-04-14' '2018-04-15' '2018-04-16' '2018-04-17' '2018-04-18'
 '2018-04-19' '2018-04-20' '2018-04-21' '2018-04-22' '2018-04-23'
 '2018-04-24' '2018-04-25' '2018-04-26' '2018-04-27' '2018-04-28'
 '2018-04-29' '2018-04-30' '2018-05-01' '2018-05-02' '2018-05-03'
 '2018-05-04' '2018-05-05' '2018-05-06' '201

### Number of Missing Values Each Column

In [ ]:
df.isnull().sum()

,0
match_id,0
date,0
season,0
event,0
venue,0
city,12397
team1,0
team2,0
toss_winner,0
toss_decision,0


### Number of Duplicate Rows

In [ ]:
df.duplicated().sum()

np.int64(0)

## 3. Data Preprocessing

3.1 Date Preprocessing

* Convert Date from Object Type to Date Time format

* Sort the data according to the date

In [ ]:
df['date'] = pd.to_datetime(df['date'])

In [ ]:
df = df.sort_values(by='date').reset_index(drop=True)
display(df.head())


,match_id,date,season,event,venue,city,team1,team2,toss_winner,toss_decision,...,non_striker,runs_batter,runs_extras,runs_total,extras_wides,extras_noballs,extras_byes,extras_legbyes,wicket_kind,wicket_player_out
0,335982,2008-04-18,2007/08,Indian Premier League,M Chinnaswamy Stadium,Bangalore,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,...,BB McCullum,0,1,1,0,0,0,1,NaN,NaN
1,335982,2008-04-18,2007/08,Indian Premier League,M Chinnaswamy Stadium,Bangalore,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,...,JH Kallis,2,0,2,0,0,0,0,NaN,NaN
2,335982,2008-04-18,2007/08,Indian Premier League,M Chinnaswamy Stadium,Bangalore,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,...,JH Kallis,1,0,1,0,0,0,0,NaN,NaN
3,335982,2008-04-18,2007/08,Indian Premier League,M Chinnaswamy Stadium,Bangalore,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,...,W Jaffer,0,0,0,0,0,0,0,NaN,NaN
4,335982,2008-04-18,2007/08,Indian Premier League,M Chinnaswamy Stadium,Bangalore,Royal Challengers Bangalore,Kolkata Knight Riders,Royal Challengers Bangalore,field,...,W Jaffer,0,1,1,0,0,0,1,NaN,NaN


3.2 Season Column Preprocessing

In [ ]:
df['season'].unique()

array(['2007/08', '2009', '2009/10', '2011', 2011, 2012, 2013, 2014, 2015,
       2016, 2017, 2018, 2019, '2019', '2020/21', '2021', 2021, 2022,
       2023, 2024, 2025, '2025', 2026], dtype=object)

* Data Type Mismatch -> The same season appears as both an integer and a string (e.g., '2001' as string and 2001 as int)


* Convert all to string data type.

In [ ]:
df['season'] = df['season'].astype(str)
df['season'].unique()

array(['2007/08', '2009', '2009/10', '2011', '2012', '2013', '2014',
       '2015', '2016', '2017', '2018', '2019', '2020/21', '2021', '2022',
       '2023', '2024', '2025', '2026'], dtype=object)

3.3 Event Preprocessing

* Since all data belongs to the same event (i.e., Indian Premier League)

* There is no use of having it. We will remove it.

In [ ]:
df = df.drop(columns=['event'])
print(df.columns)

Index(['match_id', 'date', 'season', 'venue', 'city', 'team1', 'team2',
       'toss_winner', 'toss_decision', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'innings', 'batting_team', 'over',
       'ball', 'batter', 'bowler', 'non_striker', 'runs_batter', 'runs_extras',
       'runs_total', 'extras_wides', 'extras_noballs', 'extras_byes',
       'extras_legbyes', 'wicket_kind', 'wicket_player_out'],
      dtype='object')


3.4 Venue Preprocessing

In [ ]:
df['venue'].unique()

array(['M Chinnaswamy Stadium', 'Feroz Shah Kotla',
       'Punjab Cricket Association Stadium, Mohali', 'Eden Gardens',
       'Wankhede Stadium', 'Sawai Mansingh Stadium',
       'Rajiv Gandhi International Stadium, Uppal',
       'MA Chidambaram Stadium, Chepauk', 'Dr DY Patil Sports Academy',
       'Newlands', "St George's Park", 'Kingsmead', 'SuperSport Park',
       'Buffalo Park', 'New Wanderers Stadium', 'De Beers Diamond Oval',
       'OUTsurance Oval', 'Brabourne Stadium',
       'Sardar Patel Stadium, Motera', 'Barabati Stadium',
       'Brabourne Stadium, Mumbai',
       'Vidarbha Cricket Association Stadium, Jamtha',
       'Himachal Pradesh Cricket Association Stadium', 'Nehru Stadium',
       'Holkar Cricket Stadium',
       'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium',
       'Subrata Roy Sahara Stadium',
       'Maharashtra Cricket Association Stadium',
       'Shaheed Veer Narayan Singh International Stadium',
       'JSCA International Stadium Complex', 'Sh

* Same Stadium, Different Names (e.g., "M Chinnaswamy Stadium",
"M.Chinnaswamy Stadium",
"M Chinnaswamy Stadium, Bengaluru").

* Also, we have a separate column for city, so it's better to remove the city name present in the stadium name.

In [ ]:
venue_map = {

    # Chinnaswamy Stadium
    "M.Chinnaswamy Stadium": "M Chinnaswamy Stadium",
    "M Chinnaswamy Stadium, Bengaluru": "M Chinnaswamy Stadium",

    # Arun Jaitley / Feroz Shah Kotla
    "Feroz Shah Kotla": "Arun Jaitley Stadium",
    "Arun Jaitley Stadium, Delhi": "Arun Jaitley Stadium",

    # MA Chidambaram Stadium
    "MA Chidambaram Stadium, Chepauk": "MA Chidambaram Stadium",
    "MA Chidambaram Stadium, Chepauk, Chennai": "MA Chidambaram Stadium",

    # Rajiv Gandhi International Stadium
    "Rajiv Gandhi International Stadium, Uppal": "Rajiv Gandhi International Stadium",
    "Rajiv Gandhi International Stadium, Uppal, Hyderabad": "Rajiv Gandhi International Stadium",

    # Wankhede Stadium
    "Wankhede Stadium, Mumbai": "Wankhede Stadium",

    # Punjab Cricket Association Stadium / IS Bindra Stadium
    "Punjab Cricket Association Stadium, Mohali": "Punjab Cricket Association IS Bindra Stadium",
    "Punjab Cricket Association IS Bindra Stadium, Mohali": "Punjab Cricket Association IS Bindra Stadium",
    "Punjab Cricket Association IS Bindra Stadium, Mohali, Chandigarh": "Punjab Cricket Association IS Bindra Stadium",

    # Dr DY Patil
    "Dr DY Patil Sports Academy, Mumbai": "Dr DY Patil Sports Academy",

    # ACA-VDCA Stadium
    "Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium, Visakhapatnam":
        "Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium",

    # Maharashtra Cricket Association Stadium
    "Maharashtra Cricket Association Stadium, Pune":
        "Maharashtra Cricket Association Stadium",

    # Eden Gardens
    "Eden Gardens, Kolkata": "Eden Gardens",


    # Modi Stadium
    "Sardar Patel Stadium, Motera": "Narendra Modi Stadium",

    "Narendra Modi Stadium, Ahmedabad": "Narendra Modi Stadium",

    # Sheikh Zayed Stadium
    "Zayed Cricket Stadium, Abu Dhabi": "Sheikh Zayed Stadium",

    # HPCA Stadium
    "Himachal Pradesh Cricket Association Stadium, Dharamsala":
        "Himachal Pradesh Cricket Association Stadium",

    # Sawai Mansingh Stadium
    "Sawai Mansingh Stadium, Jaipur": "Sawai Mansingh Stadium",

    # Ekana Stadium
    "Bharat Ratna Shri Atal Bihari Vajpayee Ekana Cricket Stadium, Lucknow":
        "Ekana Cricket Stadium",

    # Maharaja Yadavindra Singh Stadium
    "Maharaja Yadavindra Singh International Cricket Stadium, Mullanpur":
        "Maharaja Yadavindra Singh International Cricket Stadium",

    "Maharaja Yadavindra Singh International Cricket Stadium, New Chandigarh":
        "Maharaja Yadavindra Singh International Cricket Stadium",

    #Brabourne Stadium
    "Brabourne Stadium, Mumbai": "Brabourne Stadium",

    #Barsapara Cricket Stadium
    "Barsapara Cricket Stadium, Guwahati": "Barsapara Cricket Stadium",

    #Vidarbha Cricket Association Stadium
    "Vidarbha Cricket Association Stadium, Jamtha":
        "Vidarbha Cricket Association Stadium",
}

In [ ]:
df["venue"] = df["venue"].replace(venue_map)
df["venue"].unique()

array(['M Chinnaswamy Stadium', 'Arun Jaitley Stadium',
       'Punjab Cricket Association IS Bindra Stadium', 'Eden Gardens',
       'Wankhede Stadium', 'Sawai Mansingh Stadium',
       'Rajiv Gandhi International Stadium', 'MA Chidambaram Stadium',
       'Dr DY Patil Sports Academy', 'Newlands', "St George's Park",
       'Kingsmead', 'SuperSport Park', 'Buffalo Park',
       'New Wanderers Stadium', 'De Beers Diamond Oval',
       'OUTsurance Oval', 'Brabourne Stadium', 'Narendra Modi Stadium',
       'Barabati Stadium', 'Vidarbha Cricket Association Stadium',
       'Himachal Pradesh Cricket Association Stadium', 'Nehru Stadium',
       'Holkar Cricket Stadium',
       'Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium',
       'Subrata Roy Sahara Stadium',
       'Maharashtra Cricket Association Stadium',
       'Shaheed Veer Narayan Singh International Stadium',
       'JSCA International Stadium Complex', 'Sheikh Zayed Stadium',
       'Sharjah Cricket Stadium', 'Dubai Intern

3.5 City Preprocessing

In [ ]:
df['city'].isnull().sum()

np.int64(12397)

* City column has 12397 missing values.
* Fill the values based on the stadium name where it is located.

In [ ]:
venue_city_map = {

    # India
    "M Chinnaswamy Stadium": "Bengaluru",
    "Arun Jaitley Stadium": "Delhi",
    "Punjab Cricket Association IS Bindra Stadium": "Mohali",
    "Eden Gardens": "Kolkata",
    "Wankhede Stadium": "Mumbai",
    "Sawai Mansingh Stadium": "Jaipur",
    "Rajiv Gandhi International Stadium": "Hyderabad",
    "MA Chidambaram Stadium": "Chennai",
    "Dr DY Patil Sports Academy": "Mumbai",
    "Brabourne Stadium": "Mumbai",
    "Narendra Modi Stadium": "Ahmedabad",
    "Barabati Stadium": "Cuttack",
    "Vidarbha Cricket Association Stadium": "Nagpur",
    "Himachal Pradesh Cricket Association Stadium": "Dharamsala",
    "Nehru Stadium": "Kochi",
    "Holkar Cricket Stadium": "Indore",
    "Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium": "Visakhapatnam",
    "Subrata Roy Sahara Stadium": "Pune",
    "Maharashtra Cricket Association Stadium": "Pune",
    "Shaheed Veer Narayan Singh International Stadium": "Raipur",
    "JSCA International Stadium Complex": "Ranchi",
    "Saurashtra Cricket Association Stadium": "Rajkot",
    "Green Park": "Kanpur",
    "Ekana Cricket Stadium": "Lucknow",
    "Barsapara Cricket Stadium": "Guwahati",
    "Maharaja Yadavindra Singh International Cricket Stadium": "Mullanpur",

    # UAE
    "Sheikh Zayed Stadium": "Abu Dhabi",
    "Sharjah Cricket Stadium": "Sharjah",
    "Dubai International Cricket Stadium": "Dubai",

    # South Africa (IPL 2009)
    "Newlands": "Cape Town",
    "St George's Park": "Gqeberha",
    "Kingsmead": "Durban",
    "SuperSport Park": "Centurion",
    "Buffalo Park": "East London",
    "New Wanderers Stadium": "Johannesburg",
    "De Beers Diamond Oval": "Kimberley",
    "OUTsurance Oval": "Bloemfontein"
}

In [ ]:
df["city"] = df["city"].fillna(df["venue"].map(venue_city_map))

In [ ]:
df['city'].isnull().sum()

np.int64(0)

In [ ]:
df['city'].unique()

array(['Bangalore', 'Delhi', 'Chandigarh', 'Kolkata', 'Mumbai', 'Jaipur',
       'Hyderabad', 'Chennai', 'Cape Town', 'Port Elizabeth', 'Durban',
       'Centurion', 'East London', 'Johannesburg', 'Kimberley',
       'Bloemfontein', 'Ahmedabad', 'Cuttack', 'Nagpur', 'Dharamsala',
       'Kochi', 'Indore', 'Visakhapatnam', 'Pune', 'Raipur', 'Ranchi',
       'Abu Dhabi', 'Sharjah', 'Dubai', 'Rajkot', 'Kanpur', 'Bengaluru',
       'Navi Mumbai', 'Lucknow', 'Guwahati', 'Mohali', 'New Chandigarh'],
      dtype=object)

* Same city with different names (e.g., Bangalore, Bengaluru)

* The stadium is located in Mohali (Punjab), although older IPL datasets often use Chandigarh.

In [ ]:
city_map = {
    "Bangalore": "Bengaluru",
    "Chandigarh": "Mohali"
}

In [ ]:
df["city"] = df["city"].replace(city_map)

In [ ]:
df['city'].unique()

array(['Bengaluru', 'Delhi', 'Mohali', 'Kolkata', 'Mumbai', 'Jaipur',
       'Hyderabad', 'Chennai', 'Cape Town', 'Port Elizabeth', 'Durban',
       'Centurion', 'East London', 'Johannesburg', 'Kimberley',
       'Bloemfontein', 'Ahmedabad', 'Cuttack', 'Nagpur', 'Dharamsala',
       'Kochi', 'Indore', 'Visakhapatnam', 'Pune', 'Raipur', 'Ranchi',
       'Abu Dhabi', 'Sharjah', 'Dubai', 'Rajkot', 'Kanpur', 'Navi Mumbai',
       'Lucknow', 'Guwahati', 'New Chandigarh'], dtype=object)

3.6 Team Name Preprocessing

In [ ]:
df['team1'].unique()

array(['Royal Challengers Bangalore', 'Delhi Daredevils',
       'Kings XI Punjab', 'Kolkata Knight Riders', 'Mumbai Indians',
       'Rajasthan Royals', 'Deccan Chargers', 'Chennai Super Kings',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Gujarat Lions', 'Rising Pune Supergiants',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Lucknow Super Giants', 'Gujarat Titans',
       'Royal Challengers Bengaluru'], dtype=object)

In [ ]:
df['team2'].unique()

array(['Kolkata Knight Riders', 'Rajasthan Royals', 'Chennai Super Kings',
       'Deccan Chargers', 'Royal Challengers Bangalore',
       'Kings XI Punjab', 'Delhi Daredevils', 'Mumbai Indians',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Rising Pune Supergiants', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Gujarat Titans', 'Lucknow Super Giants',
       'Royal Challengers Bengaluru'], dtype=object)

In [ ]:
df['toss_winner'].unique()

array(['Royal Challengers Bangalore', 'Rajasthan Royals',
       'Chennai Super Kings', 'Deccan Chargers', 'Mumbai Indians',
       'Kings XI Punjab', 'Kolkata Knight Riders', 'Delhi Daredevils',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Gujarat Lions', 'Rising Pune Supergiants',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Gujarat Titans', 'Lucknow Super Giants',
       'Royal Challengers Bengaluru'], dtype=object)

In [ ]:
df['winner'].unique()

array(['Kolkata Knight Riders', 'Delhi Daredevils', 'Chennai Super Kings',
       'Royal Challengers Bangalore', 'Rajasthan Royals',
       'Kings XI Punjab', 'Deccan Chargers', 'Mumbai Indians', 'tie',
       'Pune Warriors', 'Kochi Tuskers Kerala', 'no result',
       'Sunrisers Hyderabad', 'Rising Pune Supergiants', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Gujarat Titans', 'Lucknow Super Giants',
       'Royal Challengers Bengaluru'], dtype=object)

In [ ]:
df['batting_team'].unique()

array(['Kolkata Knight Riders', 'Royal Challengers Bangalore',
       'Rajasthan Royals', 'Kings XI Punjab', 'Delhi Daredevils',
       'Chennai Super Kings', 'Deccan Chargers', 'Mumbai Indians',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Rising Pune Supergiants', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Gujarat Titans', 'Lucknow Super Giants',
       'Royal Challengers Bengaluru'], dtype=object)

* Franchise Name Inconsistency

In [ ]:
team_map = {
    "Royal Challengers Bangalore": "Royal Challengers Bengaluru",
    "Delhi Daredevils": "Delhi Capitals",
    "Kings XI Punjab": "Punjab Kings",
    "Rising Pune Supergiants": "Rising Pune Supergiant"
}

In [ ]:
cols = ['team1', 'team2', 'toss_winner', 'winner', 'batting_team']

for col in cols:
    df[col] = df[col].replace(team_map)

3.7 Toss Decision Preprocessing

* No preprocessing required because there are no null values and no data inconsistency.

3.8 Player Name Preprocessing

In [ ]:
columns_to_check = ['player_of_match', 'batter', 'bowler', 'non_striker', 'wicket_kind', 'wicket_player_out']

for col in columns_to_check:
    print(f"\nUnique values in '{col}':")
    print(df[col].unique())


Unique values in 'player_of_match':
['BB McCullum' 'MF Maharoof' 'MEK Hussey' 'DJ Hussey' 'MV Boucher'
 'SR Watson' 'V Sehwag' 'ML Hayden' 'YK Pathan' 'KC Sangakkara' 'JDP Oram'
 'SM Katich' 'AC Gilchrist' 'MS Dhoni' 'ST Jayasuriya' 'GD McGrath'
 'SA Asnodkar' 'SE Marsh' 'IK Pathan' 'P Kumar' 'Sohail Tanvir'
 'SM Pollock' 'S Sreesanth' 'A Nehra' 'SC Ganguly' 'L Balaji'
 'Shoaib Akhtar' 'A Mishra' 'DPMD Jayawardene' 'GC Smith' 'DJ Bravo'
 'M Ntini' 'SP Goswami' 'A Kumble' 'KD Karthik' 'JA Morkel'
 'R Vinay Kumar' 'Umar Gul' 'SK Raina' 'CRD Fernando' 'R Dravid'
 'SR Tendulkar' 'RP Singh' 'DL Vettori' 'M Muralitharan' 'CH Gayle'
 'AB de Villiers' 'RS Bopara' 'PP Ojha' 'TM Dilshan' 'HH Gibbs'
 'DP Nannes' 'Yuvraj Singh' 'JP Duminy' 'SB Jakati' 'JH Kallis'
 'G Gambhir' 'RG Sharma' 'A Singh' 'S Badrinath' 'DR Smith'
 'Harbhajan Singh' 'LRPL Taylor' 'R Bhatia' 'SK Warne' 'B Lee' 'BJ Hodge'
 'LR Shukla' 'MK Pandey' 'AD Mathews' 'WPUJC Vaas' 'MK Tiwary' 'A Symonds'
 'AA Jhunjhunwala' 'J Theron

In [ ]:
columns_to_check_nulls = ['player_of_match', 'batter', 'bowler', 'non_striker', 'wicket_kind', 'wicket_player_out']

for col in columns_to_check_nulls:
    print(f"\nNumber of null values in '{col}': {df[col].isnull().sum()}")


Number of null values in 'player_of_match': 828

Number of null values in 'batter': 0

Number of null values in 'bowler': 0

Number of null values in 'non_striker': 0

Number of null values in 'wicket_kind': 275267

Number of null values in 'wicket_player_out': 275267


In [ ]:
df['player_of_match']=df['player_of_match'].fillna("No Award", inplace=True)
df['wicket_kind'] = df['wicket_kind'].fillna('No Wicket')
df['wicket_player_out'] = df['wicket_player_out'].fillna('No Dismissal')

/tmp/ipykernel_1291/754778715.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['player_of_match']=df['player_of_match'].fillna("No Award", inplace=True)


3.9 Numerical Value Preprocessing

* No preprocessing required.
* Keep NaN values as is.

In [ ]:
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns

for col in numerical_cols:
    print(f"\nUnique values in numerical column '{col}':")
    print(df[col].unique())


Unique values in numerical column 'match_id':
[ 335982  335984  335983 ... 1529290 1529291 1529292]

Unique values in numerical column 'win_by_runs':
[140.  nan  33.   6.  66.  13.  10.  45.   9.   3.  29.   5.  18.  23.
  12.  65.  25.   1.  14.  41. 105.  75.  19.  92.  11.  24.  27.  38.
   8.  78.  16.  53.   2.   4.  31.  55.  98.  34.  36.  17.  39.  40.
  67.  63.  37.  57.  35.  22.  21.  48.  26.  20.  85.  32.  76. 111.
  82.  43.  58.  28.  74.  42.  59.  46.  47.   7.  86.  44.  87. 130.
  15.  60.  77.  30.  50.  93.  72.  62.  97. 138.  71. 144.  80.  51.
  61. 146.  64. 102. 118.  49.  69.  88.  54.  91.  52.  81.  56. 112.
 106. 100. 110.  83.  99. 103.]

Unique values in numerical column 'win_by_wickets':
[nan  9.  5.  6.  3.  7.  4. 10.  8.  2.  1.]

Unique values in numerical column 'innings':
[1 2 4 3 5 6]

Unique values in numerical column 'over':
[ 0  3  4  2  5  6 18 19  1  7 11 12 13 14 15  8  9 10 16 17]

Unique values in numerical column 'ball':
[ 1  2  3  4 

-----------------------------------------------OVER-------------------------------------------------------------
